File: sdk/typescript/src/agents.ts
/**
 * Agent admin API (v1.5).
 *
 * Register, list, and revoke (deactivate) OAuth agent clients.
 *
 * Routes (all admin-key authenticated):
 *   POST   /api/v1/agents                        — create / register agent
 *   GET    /api/v1/agents                        — list agents
 *   DELETE /api/v1/agents/{id}                   — deactivate (revoke) agent
 *   GET    /api/v1/agents/{id}                   — get agent by ID or clientID
 *   GET    /api/v1/agents/{id}/tokens            — list active tokens
 *   POST   /api/v1/agents/{id}/tokens/revoke-all — revoke all tokens
 *   POST   /api/v1/agents/{id}/rotate-secret     — rotate client secret
 *   POST   /api/v1/agents/{id}/rotate-dpop-key   — rotate DPoP keypair
 */

import { httpRequest } from "./http.js";
import { SharkAPIError } from "./errors.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** A registered agent OAuth client. */
export interface Agent {
  id: string;
  name: string;
  description?: string;
  client_id: string;
  client_type: string;
  auth_method: string;
  redirect_uris: string[];
  grant_types: string[];
  response_types: string[];
  scopes: string[];
  token_lifetime: number;
  metadata: Record<string, unknown>;
  logo_uri?: string;
  homepage_uri?: string;
  active: boolean;
  created_at: string;
  updated_at: string;
}

/**
 * Response from `registerAgent`. Includes the one-time `client_secret`.
 * The secret is not stored server-side (only its hash is kept) — copy it now.
 */
export interface AgentCreateResult extends Agent {
  /** Plaintext client secret. Shown exactly once. */
  client_secret: string;
}

/** Input for registering a new agent. */
export interface RegisterAgentInput {
  /** Human-readable label. Required. */
  name: string;
  description?: string;
  /** `"confidential"` (default) or `"public"`. */
  client_type?: string;
  /** `"client_secret_basic"` (default). */
  auth_method?: string;
  redirect_uris?: string[];
  grant_types?: string[];
  response_types?: string[];
  /** OAuth scopes the agent is allowed to request. */
  scopes?: string[];
  /** Token lifetime in seconds. Default: 3600. */
  token_lifetime?: number;
  metadata?: Record<string, unknown>;
  logo_uri?: string;
  homepage_uri?: string;
}

/** A single active token record for an agent. */
export interface TokenInfo {
  token_id: string;
  agent_id: string;
  jkt?: string;
  scope?: string;
  expires_at?: string;
  created_at?: string;
}

/** Result of a revoke-all tokens operation. */
export interface RevokeAllResult {
  revoked_count: number;
  agent_id: string;
}

/** New credentials returned after a secret rotation. */
export interface AgentCredentials {
  agent_id: string;
  client_id: string;
  /** Plaintext client secret — copy now, shown once. */
  client_secret: string;
  rotated_at?: string;
}

/** Result of a DPoP key rotation operation. */
export interface DPoPRotationResult {
  old_jkt: string;
  new_jkt: string;
  revoked_token_count: number;
  audit_event_id: string;
}

/** Options for `listAgents`. */
export interface ListAgentsOptions {
  /** Filter by active/inactive state. Omit for all. */
  active?: boolean;
  /** Search term matched against name. */
  search?: string;
  limit?: number;
  offset?: number;
}

/** Response from `listAgents`. */
export interface AgentListResult {
  data: Agent[];
  total: number;
}

// ---------------------------------------------------------------------------
// Client options
// ---------------------------------------------------------------------------

/** Options for {@link AgentsClient}. */
export interface AgentsClientOptions {
  /** Base URL of the SharkAuth server. */
  baseUrl: string;
  /** Admin API key (Bearer token). */
  adminKey: string;
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

/**
 * Admin client for managing agent OAuth clients.
 */
export class AgentsClient {
  private readonly _base: string;
  private readonly _key: string;

  constructor(opts: AgentsClientOptions) {
    this._base = opts.baseUrl.replace(/\/+$/, "");
    this._key = opts.adminKey;
  }

  private _auth(): Record<string, string> {
    return { Authorization: `Bearer ${this._key}` };
  }

  private async _throw(status: number, text: string): Promise<never> {
    let code = "api_error";
    let message = text.slice(0, 300);
    try {
      const body = JSON.parse(text) as {
        error?: { code?: string; message?: string };
      };
      if (body.error?.code) code = body.error.code;
      if (body.error?.message) message = body.error.message;
    } catch {
      // keep defaults
    }
    throw new SharkAPIError(message, code, status);
  }

  /**
   * Register (create) a new agent client.
   *
   * The returned `client_secret` is shown exactly once — store it securely.
   */
  async registerAgent(input: RegisterAgentInput): Promise<AgentCreateResult> {
    const url = `${this._base}/api/v1/agents`;
    const resp = await httpRequest(url, {
      method: "POST",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify(input),
    });
    if (resp.status !== 201) await this._throw(resp.status, resp.text);
    return resp.json<AgentCreateResult>();
  }

  /** List agents with optional filters. */
  async listAgents(opts?: ListAgentsOptions): Promise<AgentListResult> {
    const qs = new URLSearchParams();
    if (opts?.active != null) qs.set("active", String(opts.active));
    if (opts?.search) qs.set("search", opts.search);
    if (opts?.limit != null) qs.set("limit", String(opts.limit));
    if (opts?.offset != null) qs.set("offset", String(opts.offset));
    const query = qs.toString();
    const url = `${this._base}/api/v1/agents${query ? `?${query}` : ""}`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    return resp.json<AgentListResult>();
  }

  /** Get a single agent by ID or clientID. */
  async getAgent(id: string): Promise<Agent> {
    const url = `${this._base}/api/v1/agents/${encodeURIComponent(id)}`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    return resp.json<Agent>();
  }

  /**
   * Deactivate (revoke) an agent by ID.
   *
   * This sets `active=false` and revokes all outstanding OAuth tokens for the
   * agent's `client_id`. Returns void on 204.
   */
  async revokeAgent(id: string): Promise<void> {
    const url = `${this._base}/api/v1/agents/${encodeURIComponent(id)}`;
    const resp = await httpRequest(url, {
      method: "DELETE",
      headers: this._auth(),
    });
    if (resp.status !== 204) await this._throw(resp.status, resp.text);
  }

  // ------------------------------------------------------------------
  // Token management
  // ------------------------------------------------------------------

  /** List active tokens for an agent. GET /api/v1/agents/{id}/tokens */
  async listTokens(agentId: string): Promise<TokenInfo[]> {
    const url = `${this._base}/api/v1/agents/${encodeURIComponent(agentId)}/tokens`;
    const resp = await httpRequest(url, { headers: this._auth() });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    const body = resp.json<{ data?: TokenInfo[] } | TokenInfo[]>();
    const items = Array.isArray(body) ? body : (body as { data?: TokenInfo[] }).data ?? [];
    return items.map((t) => ({
      token_id: (t as Record<string, string>).id ?? t.token_id,
      agent_id: t.agent_id ?? agentId,
      jkt: t.jkt,
      scope: t.scope,
      expires_at: t.expires_at,
      created_at: t.created_at,
    }));
  }

  /** Revoke all active tokens for an agent. POST /api/v1/agents/{id}/tokens/revoke-all */
  async revokeAllTokens(agentId: string): Promise<RevokeAllResult> {
    const url = `${this._base}/api/v1/agents/${encodeURIComponent(agentId)}/tokens/revoke-all`;
    const resp = await httpRequest(url, { method: "POST", headers: this._auth() });
    if (resp.status !== 200 && resp.status !== 204) await this._throw(resp.status, resp.text);
    if (resp.status === 204 || !resp.text) {
      return { revoked_count: 0, agent_id: agentId };
    }
    const data = resp.json<{ revoked_count?: number; count?: number }>();
    return { revoked_count: data.revoked_count ?? data.count ?? 0, agent_id: agentId };
  }

  /** Rotate the agent's client secret. POST /api/v1/agents/{id}/rotate-secret */
  async rotateSecret(agentId: string): Promise<AgentCredentials> {
    const url = `${this._base}/api/v1/agents/${encodeURIComponent(agentId)}/rotate-secret`;
    const resp = await httpRequest(url, { method: "POST", headers: this._auth() });
    if (resp.status !== 200 && resp.status !== 201) await this._throw(resp.status, resp.text);
    const raw = resp.json<{ data?: AgentCredentials } & AgentCredentials>();
    const data = raw.data ?? raw;
    return {
      agent_id: data.agent_id ?? agentId,
      client_id: data.client_id ?? "",
      client_secret: data.client_secret ?? "",
      rotated_at: data.rotated_at,
    };
  }

  /**
   * Rotate an agent's DPoP keypair binding (admin-forced).
   * POST /api/v1/agents/{id}/rotate-dpop-key
   *
   * @param agentId         The `agent_*` identifier.
   * @param newPublicKeyJwk New public key as JWK. Server derives `cnf.jkt` via RFC 7638 SHA-256.
   * @param reason          Optional audit-log reason string.
   */
  async rotateDpopKey(
    agentId: string,
    newPublicKeyJwk: Record<string, unknown>,
    reason?: string,
  ): Promise<DPoPRotationResult> {
    const url = `${this._base}/api/v1/agents/${encodeURIComponent(agentId)}/rotate-dpop-key`;
    const body: Record<string, unknown> = { new_public_jwk: newPublicKeyJwk };
    if (reason !== undefined) body.reason = reason;
    const resp = await httpRequest(url, {
      method: "POST",
      headers: { ...this._auth(), "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });
    if (resp.status !== 200) await this._throw(resp.status, resp.text);
    const data = resp.json<DPoPRotationResult>();
    return {
      old_jkt: data.old_jkt ?? "",
      new_jkt: data.new_jkt ?? "",
      revoked_token_count: data.revoked_token_count ?? 0,
      audit_event_id: data.audit_event_id ?? "",
    };
  }
}

File: sdk/python/shark_auth/agents.py
"""Agent management — admin API."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Dict, List, Optional

from . import _http
from .proxy_rules import _raise


# ---------------------------------------------------------------------------
# W2 Method 5 — typed response dataclasses
# ---------------------------------------------------------------------------


@dataclass
class TokenInfo:
    """A single active token record for an agent.

    Attributes
    ----------
    token_id:
        Opaque token identifier.
    agent_id:
        The agent this token belongs to.
    jkt:
        DPoP JWK thumbprint, if the token is DPoP-bound.
    scope:
        Space-delimited scopes granted.
    expires_at:
        ISO-8601 expiry timestamp.
    created_at:
        ISO-8601 creation timestamp.
    """

    token_id: str
    agent_id: str
    jkt: Optional[str] = None
    scope: Optional[str] = None
    expires_at: Optional[str] = None
    created_at: Optional[str] = None


@dataclass
class RevokeResult:
    """Result of a revoke-all operation.

    Attributes
    ----------
    revoked_count:
        Number of tokens revoked.
    agent_id:
        The agent whose tokens were revoked.
    """

    revoked_count: int
    agent_id: str


@dataclass
class AgentCredentials:
    """New credentials returned after a secret rotation.

    Attributes
    ----------
    agent_id:
        The agent whose secret was rotated.
    client_id:
        OAuth client ID (unchanged).
    client_secret:
        New client secret — copy now, will not be shown again.
    rotated_at:
        ISO-8601 timestamp of the rotation.
    """

    agent_id: str
    client_id: str
    client_secret: str
    rotated_at: Optional[str] = None


@dataclass
class AuditEvent:
    """A single audit-log entry.

    Attributes
    ----------
    id:
        Unique audit event identifier.
    event:
        Event type string (e.g. ``"agent.token_issued"``).
    actor_id:
        ID of the principal that triggered the event.
    target_id:
        ID of the affected resource, if applicable.
    metadata:
        Arbitrary event metadata dict.
    created_at:
        ISO-8601 timestamp.
    """

    id: str
    event: str
    actor_id: Optional[str] = None
    target_id: Optional[str] = None
    metadata: Dict[str, Any] = field(default_factory=dict)
    created_at: Optional[str] = None


class AgentsClient:
    """Admin client for managing machine-to-machine agents.

    Agents correspond to OAuth 2.1 clients registered in SharkAuth.
    This client wraps the ``/api/v1/agents`` admin routes.

    Parameters
    ----------
    base_url:
        Base URL of the SharkAuth server.
    token:
        Admin API key (``sk_live_...``).
    session:
        Optional pre-configured :class:`requests.Session`.
    """

    _PREFIX = "/api/v1/agents"

    def __init__(
        self,
        base_url: str,
        token: str,
        *,
        session: Optional[object] = None,
    ) -> None:
        self._base = base_url.rstrip("/")
        self._token = token
        self._session = session or _http.new_session()

    def _auth(self) -> Dict[str, str]:
        return {"Authorization": f"Bearer {self._token}"}

    # ------------------------------------------------------------------
    # Register (create)
    # ------------------------------------------------------------------

    def register_agent(
        self,
        app_id: str,
        name: str,
        scopes: Optional[List[str]] = None,
        **extra: Any,
    ) -> Dict[str, Any]:
        """Register a new agent and return the created agent object.

        The response includes a one-time ``client_secret`` — copy it now as
        the server will not return it again.

        Parameters
        ----------
        app_id:
            Application ID to scope the agent to (stored in ``metadata``).
        name:
            Human-readable agent name (required by the server).
        scopes:
            List of OAuth scopes to grant.  Defaults to ``[]``.
        **extra:
            Additional fields forwarded verbatim (e.g. ``description``,
            ``token_lifetime``, ``redirect_uris``).
        """
        body: Dict[str, Any] = {
            "name": name,
            "scopes": scopes or [],
            "metadata": {"app_id": app_id},
            **extra,
        }
        url = f"{self._base}{self._PREFIX}"
        resp = _http.request(self._session, "POST", url, headers=self._auth(), json=body)
        if resp.status_code in (200, 201):
            return resp.json()
        _raise(resp)

    # ------------------------------------------------------------------
    # List
    # ------------------------------------------------------------------

    def list_agents(self, app_id: Optional[str] = None) -> List[Dict[str, Any]]:
        """List all agents.  Filter by *app_id* when supplied."""
        params: Dict[str, str] = {}
        if app_id is not None:
            params["search"] = app_id  # server supports ?search= for name/id lookup
        url = f"{self._base}{self._PREFIX}"
        resp = _http.request(self._session, "GET", url, headers=self._auth(), params=params)
        if resp.status_code == 200:
            return resp.json().get("data", [])
        _raise(resp)

    # ------------------------------------------------------------------
    # Get single agent
    # ------------------------------------------------------------------

    def get_agent(self, agent_id: str) -> Dict[str, Any]:
        """Return a single agent by *agent_id* (or client_id).

        Parameters
        ----------
        agent_id:
            The ``id`` (``agent_*``) or ``client_id`` (``shark_agent_*``) of
            the agent to fetch.

        Example
        -------
        >>> client = AgentsClient(base_url="https://auth.example.com", token="sk_live_...")
        >>> agent = client.get_agent("agent_abc123")
        >>> print(agent["name"])
        """
        url = f"{self._base}{self._PREFIX}/{agent_id}"
        resp = _http.request(self._session, "GET", url, headers=self._auth())
        if resp.status_code == 200:
            return resp.json()
        _raise(resp)

    # ------------------------------------------------------------------
    # Revoke (deactivate)
    # ------------------------------------------------------------------

    def revoke_agent(self, agent_id: str) -> None:
        """Deactivate an agent and revoke all of its tokens.

        The server returns 204 on success.
        """
        url = f"{self._base}{self._PREFIX}/{agent_id}"
        resp = _http.request(self._session, "DELETE", url, headers=self._auth())
        if resp.status_code == 204:
            return
        _raise(resp)

    # ------------------------------------------------------------------
    # W2 Method 5 — token management extras
    # ------------------------------------------------------------------

    def list_tokens(self, agent_id: str) -> List[TokenInfo]:
        """List active tokens for an agent.

        Wraps ``GET /api/v1/agents/{id}/tokens``.

        Parameters
        ----------
        agent_id:
            The ``agent_*`` identifier of the agent.

        Returns
        -------
        list[TokenInfo]
            Active token records for the agent.

        Example
        -------
        >>> tokens = client.agents.list_tokens("agent_abc")
        >>> for tok in tokens:
        ...     print(tok.token_id, tok.scope)
        """
        url = f"{self._base}{self._PREFIX}/{agent_id}/tokens"
        resp = _http.request(self._session, "GET", url, headers=self._auth())
        if resp.status_code == 200:
            data = resp.json()
            items = data.get("data", data) if isinstance(data, dict) else data
            return [
                TokenInfo(
                    token_id=t.get("id", t.get("token_id", "")),
                    agent_id=t.get("agent_id", agent_id),
                    jkt=t.get("jkt"),
                    scope=t.get("scope"),
                    expires_at=t.get("expires_at"),
                    created_at=t.get("created_at"),
                )
                for t in (items if isinstance(items, list) else [])
            ]
        _raise(resp)

    def revoke_all(self, agent_id: str) -> RevokeResult:
        """Revoke all active tokens for an agent.

        Wraps ``POST /api/v1/agents/{id}/tokens/revoke-all``.

        Parameters
        ----------
        agent_id:
            The ``agent_*`` identifier of the agent.

        Returns
        -------
        RevokeResult
            The number of tokens revoked.

        Example
        -------
        >>> result = client.agents.revoke_all("agent_abc")
        >>> print(result.revoked_count)
        """
        url = f"{self._base}{self._PREFIX}/{agent_id}/tokens/revoke-all"
        resp = _http.request(self._session, "POST", url, headers=self._auth())
        if resp.status_code in (200, 204):
            if resp.status_code == 204 or not resp.content:
                return RevokeResult(revoked_count=0, agent_id=agent_id)
            data = resp.json()
            return RevokeResult(
                revoked_count=data.get("revoked_count", data.get("count", 0)),
                agent_id=agent_id,
            )
        _raise(resp)

    def rotate_secret(self, agent_id: str) -> AgentCredentials:
        """Rotate the agent's client secret.

        Wraps ``POST /api/v1/agents/{id}/rotate-secret``.

        Parameters
        ----------
        agent_id:
            The ``agent_*`` identifier of the agent.

        Returns
        -------
        AgentCredentials
            New credentials — copy ``client_secret`` now, not shown again.

        Example
        -------
        >>> creds = client.agents.rotate_secret("agent_abc")
        >>> print(creds.client_secret)
        """
        url = f"{self._base}{self._PREFIX}/{agent_id}/rotate-secret"
        resp = _http.request(self._session, "POST", url, headers=self._auth())
        if resp.status_code in (200, 201):
            data = resp.json()
            # Server may nest under "data" key
            if isinstance(data, dict) and "data" in data:
                data = data["data"]
            return AgentCredentials(
                agent_id=data.get("agent_id", agent_id),
                client_id=data.get("client_id", ""),
                client_secret=data.get("client_secret", ""),
                rotated_at=data.get("rotated_at"),
            )
        _raise(resp)

    def get_audit_logs(
        self,
        agent_id: str,
        *,
        limit: int = 100,
        since: Optional[datetime] = None,
    ) -> List[AuditEvent]:
        """Fetch audit events filtered to this agent.

        Wraps ``GET /api/v1/audit-logs?actor_id=<id>``.

        Parameters
        ----------
        agent_id:
            The ``agent_*`` identifier of the agent (used as ``actor_id`` filter).
        limit:
            Maximum number of events to return. Default: 100.
        since:
            Only return events after this datetime. Default: no filter.

        Returns
        -------
        list[AuditEvent]
            Audit log entries for the agent, newest first.

        Example
        -------
        >>> events = client.agents.get_audit_logs("agent_abc", limit=10)
        >>> for ev in events:
        ...     print(ev.event, ev.created_at)
        """
        params: Dict[str, Any] = {"actor_id": agent_id, "limit": limit}
        if since is not None:
            params["since"] = since.isoformat()
        url = f"{self._base}/api/v1/audit-logs"
        resp = _http.request(self._session, "GET", url, headers=self._auth(), params=params)
        if resp.status_code == 200:
            data = resp.json()
            items = data.get("data", data) if isinstance(data, dict) else data
            return [
                AuditEvent(
                    id=e.get("id", ""),
                    event=e.get("event", e.get("action", "")),
                    actor_id=e.get("actor_id"),
                    target_id=e.get("target_id"),
                    metadata=e.get("metadata", {}),
                    created_at=e.get("created_at"),
                )
                for e in (items if isinstance(items, list) else [])
            ]
        _raise(resp)
